Add UDF doc

luna-bianca opened the pull request:
## What are you changing in this pull request and why?
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7838
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7843
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7844
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7845
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7954
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7958
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7957
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7989
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/7990

## Checklist
<!-- Ensure your PR meets the following items. Feel free to add additiona checklist items to the list if additional checks need to happen before the PR is merged, such as:
- [ ] Needs technical review
- [ ] Change base branch
- [ ] Merge on xyz date
-->
- [x] The changes in this PR meet the [docs style guide/fundamentals](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) required for all content.
- [ ] Applied the proper versioning rules if the content is for specific dbt version(s): ([version a whole page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) or [version a block of content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content)
- [ ] The content in this PR requires a dbt release note, so I added one to the [release notes page](https://docs.getdbt.com/docs/dbt-versions/dbt-cloud-release-notes).).
<!--
PRE-RELEASE VERSION OF dbt (if so, uncomment):
- [ ] Add a note to the prerelease version [Migration Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
<!-- 
ADDING OR REMOVING PAGES (if so, uncomment):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/vercel.json`
- [ ] Run link testing locally with `npm run build` to update the links that point to deleted pages
-->

<!-- vercel-deployment-preview -->
---
🚀 Deployment available! Here are the direct links to the updated files:


- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/docs/build/udfs
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/docs/dbt-versions/core-upgrade/03-upgrading-to-fusion
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/docs/dbt-versions/core-upgrade/04-upgrading-to-v1.11
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/faqs/Project/udfs-vs-macros
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/dbt_project.yml
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/function-configs
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/function-properties
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/global-configs/resource-type
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/node-selection/methods
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/project-configs/function-paths
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/resource-properties/arguments
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/resource-properties/function-arguments
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/resource-properties/returns
- https://docs-getdbt-com-git-add-udfs-doc-dbt-labs.vercel.app/reference/resource-properties/type

<!-- end-vercel-deployment-preview -->

luna-bianca committed:
Add UDF doc

luna-bianca committed:
Update udfs.md

luna-bianca committed:
Update udfs.md

luna-bianca committed:
Update udfs.md

luna-bianca committed:
Update udfs.md

matthewshaver commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,110 @@
+---
+title: "User-defined functions"
+description: "Learn how to define UDFs in your dbt projects."
```
I would add the full term to the description to help SEO 
```suggestion
description: "Learn how to add user-defined functions (UDFs) to your dbt projects."
```

matthewshaver commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,110 @@
+
+User-defined functions (UDFs) enable users to define and register custom functions within the warehouse. Like [macros](/docs/build/jinja-macros), UDFs enable reuse of code; however, unlike macros, you can define UDFs in languages other than SQL (for example, Python, Java, Scala) and you can use them in queries outside <Constant name="core" />.
+
+Note that only basic SQL UDFs are currently supported in <Constant name="core" />.
```
I'm not sure if there's a better way to phrase this, but it feels contradictory to the part that immediately precedes it, where we say that they can define UDFs in languages other than SQL.

matthewshaver commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,110 @@
+
+<Constant name="core" /> creates, updates, and renames UDFs as part of DAG execution. The UDF file is created before building the model that references it.
+
+## Defining UDFs in dbt
```
dbt is used either as a catchall for all versions of dbt or specifically for cloud platform. Is this truly a Core only feature or can it be used in dbt platform as well? If "no" to the latter, then I'd suggest
```suggestion
## Defining UDFs in dbt Core
```

Otherwise we should remove the <Constant name="core" /> mentions from the text body

matthewshaver commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,110 @@
+
+    </File>
+
+2. Define your argument, output types, properties, and configs in a corresponding YAML file.
```
To keep the formatting consistent between steps
```suggestion
2. Define your argument, output types, properties, and configs in a corresponding YAML file. For example:
```

matthewshaver commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,110 @@
+
+2. Define your argument, output types, properties, and configs in a corresponding YAML file.
+
+    For example:
```
See above
```suggestion
```

matthewshaver commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,110 @@
+
+4. Run `dbt compile`.
+
+    In the following example, the `{{ ref('is_positive_int') }}` is replaced by the UDF name `udf_db.udf_schema.is_positive_int`.
```
I'd move this up next to step 4 for consistency

matthewshaver commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,110 @@
+id: "udfs"
+---
+
+## Overview
```
We try to avoid having back-to-back headers with no text in between, especially for the opening since the title will be the lead-in. You can remove the ## Overview header and go right into the description. 
```suggestion
```

luna-bianca committed:
Update website/docs/docs/build/udfs.md

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

luna-bianca committed:
Update website/docs/docs/build/udfs.md

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

luna-bianca committed:
Update website/docs/docs/build/udfs.md

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

luna-bianca committed:
Update website/docs/docs/build/udfs.md

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

luna-bianca committed:
Update udfs.md

luna-bianca committed:
Update udfs.md

luna-bianca committed:
Update UDF doc

luna-bianca committed:
Update udfs.md

luna-bianca committed:
Update udfs.md

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,152 @@
+
+- BigQuery
+- Snowflake
+- Redshift
```
And postgres! Though whether that is important to document 🤷🏻 don't know how many people use it in cloud

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,152 @@
+
+# User-defined functions <Lifecycle status="beta" />
+
+<!--Confirm how users can get this feature-->
+:::info Beta feature 
+If you are interested in getting access to this feature, please contact us. 
+:::
```
In cloud, it's going to be, available. It won't be behind any feature flag. People can just start using it if they so desire. In OSS they'll have to install a beta version of dbt-core (e.g. `pip install dbt-core 1.11.0b2`). Even though in cloud it isn't behind a feature flag, the feature itself is considered beta at this point in time.

luna-bianca replied:
@runleonarun We can include this in the Upgrade guide

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,152 @@
+            database: udf_db
+        arguments: 
+            - name: a_string
+              type: string
+              description: The string that I want to check if it's representing a positive integer (like "10") 
+        return_type: # required
+            type: boolean # required
```
This is my fault in. I originally implemented them incorrectly, but they are now fixed
* `arguments > type` should actually be `arguments > data_type`
* `return_type` should be `returns`
* `return_type > type` should be `returns > data_type`

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,152 @@
+            schema: udf_schema
+            database: udf_db
+        arguments: 
+            - name: a_string
+              type: string
```
It should be noted that although `arguments` is not required, if arguments are specified both the `name` and `data_type` for an argument is required.

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,152 @@
+    You can use these values for the `type` property when you define a function in a YAML file.
+
+    - `scalar` - Returns a single value per row
+    - `aggregate` - Returns a single value per group, aggregating several rows
+    - `table functions` - Returns a table result
```
These two are not yet supported. We have plans to, but at this point specifying them won't do what the user expects.

luna-bianca replied:
Hid this section for now

QMalcolm commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,152 @@
+    ```sql
+    CREATE OR REPLACE FUNCTION repeat_n(s VARCHAR, n INTEGER)
+    RETURNS VARCHAR
+    IMMUTABLE
````
In redshift for now it'll be `VOLATILE` instead of `IMMUTABLE`. We don't have the ability just yet to dynamically set the volatility. Thus for the time being, redshift is hard coded to `VOLATILE` as that is the safest option to set.

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,152 @@
+    select
+
+    maybe_positive_int_column,
+        {{ function('is_positive_int') }}('maybe_positive_int_column')
```
`('maybe_positive_int_column')` shouldn't be in single quotes here. That is, what we pass in to the function _is_ the column, not the name of the column as a string. That is, it should be `(maybe_positive_into_column)` instead.

QMalcolm reviewed (commented):
Thank you so much for doing this work ❤️  Some comments that we should address, but overall this is looking wonderful 🚀

luna-bianca committed:
Apply comments from review

luna-bianca commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,155 @@
+
+
+:::info
+UDFs are not yet supported in the dbt Fusion engine.
```
Added this note

QMalcolm replied:
_Yet_ is the keyword here, we're hoping to have this in by Monday though 🤞🏻

QMalcolm replied:
Fusion support is in canary. It should be available tomorrow to Fusion folks 🙂

luna-bianca commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,155 @@
+UDFs are not yet supported in the dbt Fusion engine.
+:::
+
+User-defined functions (UDFs) enable users to define and register custom functions within the warehouse. Like [macros](/docs/build/jinja-macros), UDFs enable reuse of code; however, unlike macros, you can define UDFs in languages other than SQL (for example, Python, Java, Scala) and you can use them in queries outside dbt.
```
Is defining UDFs in other languages now supported?

QMalcolm replied:
Other languages are not supported at this point. We plan on adding python support, but that'll be a follow-up _after_ coalesce 🙂

luna-bianca replied:
Revised this paragraph [here](https://github.com/dbt-labs/docs.getdbt.com/pull/7841/commits/3217713c0fd75f8fc87e2ab4b0522dec567e557a) to avoid confusion

luna-bianca committed:
Update udfs.md

luna-bianca committed:
Add more info + section about unit tests

luna-bianca commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,190 @@
+UDFs are not yet supported in the dbt Fusion engine.
+:::
+
+User-defined functions (UDFs) enable users to define and register custom functions in your warehouse. Like [macros](/docs/build/jinja-macros), UDFs promote code reuse. They run natively in the warehouse so you can reuse the same logic in tools outside dbt. Note that creating UDFs in other languages (for example, Python, Java, or Scala) is not yet supported. 
```
Revised this paragraph

luna-bianca commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,190 @@
+
+    In your DAG, there should be a dependency between `is_positive_int` → `my_model` and a UDF node is created from the SQL and YAML definitions.
+
+After defining a UDF, you can update the SQL file that contains its function body (`is_positive_int.sql` in this example). When you rebuild, your changes will be applied everywhere the UDF is referenced. 
```
Added this sentence

luna-bianca commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,190 @@
+
+After defining a UDF, you can update the SQL file that contains its function body (`is_positive_int.sql` in this example). When you rebuild, your changes will be applied everywhere the UDF is referenced. 
+
+## Using UDFs in unit tests
```
Added this section

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,190 @@
+
+dbt creates, updates, and renames UDFs as part of DAG execution. The UDF file is created before building the model that references it.
+
+Only `scalar` functions are currently supported. The `scalar` UDF type returns a single value per row.
```
@luna-bianca  I wonder if we should have a limitations or "prerequisites" section to call out:
- Only scalar functions are currently supported (returns a single value per row)
- Creating UDFs in other languages (Python, Java, Scala) is not yet supported
- list of supported adapters

luna-bianca replied:
Added a new Limitation section 👍

luna-bianca committed:
Add limitations section

luna-bianca committed:
Merge branch 'current' into add-UDFs-doc

luna-bianca committed:
Merge branch 'add-UDFs-doc' of github.com:dbt-labs/docs.getdbt.com into add-UDFs-doc

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,191 @@
+
+dbt creates, updates, and renames UDFs as part of DAG execution. The UDF file is created before building the model that references it.
+
+## Supported adapters
```
@luna-bianca @QMalcolm Are UDFs backwards compatible to previous Core versions? If not, we might want to add as a prereq.

QMalcolm replied:
UDFs are _not_ backwards compatible. That is in OSS, it's only available for `1.11+`. In cloud it's in beta for `latest`, it'll come to `compatible` and `extended` at a later date. _Defining_ UDFs _is_ backwards compatible as far as doing so won't _break_ earlier versions (they'll get ignored on parsing). However, using the `function` macro on earlier versions _will_ cause an error on earlier versions.

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,191 @@
+
+UDFs are supported in the following adapters:
+
+- BigQuery
```
We might add that it's on Core for local and dbt platform for all of these adapters. Maybe there's a better way than listing under each adapter?
```suggestion
- BigQuery 
  - <Constant name="core" /> local install
  - <Constant name="core" /> in the <Constant name="dbt_platform" />
```

luna-bianca replied:
Addressed here: https://github.com/dbt-labs/docs.getdbt.com/pull/7841/commits/2fba50931c7b40428ee2082b66bcdd6eb4b1a4a3

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,191 @@
+- BigQuery
+- Snowflake
+- Redshift
+- Postgres (<Constant name="core" /> only)
```
```suggestion
- Postgres
  - <Constant name="core" /> local install only
```

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,191 @@
+UDFs are not yet supported in the dbt Fusion engine.
+:::
+
+User-defined functions (UDFs) enable users to define and register custom functions in your warehouse. Like [macros](/docs/build/jinja-macros), UDFs promote code reuse. They run natively in the warehouse so you can reuse the same logic in tools outside dbt.
```
Changing user to you and wonder if we can add some more of the business value here. cc @QMalcolm @graciegoheen to check my technical accuracy.
```suggestion
User-defined functions (UDFs) enable you to define and register custom functions in your warehouse. Like [macros](/docs/build/jinja-macros), UDFs promote code reuse, but they are objects in the warehouse so you can reuse the same logic in tools outside dbt, such as BI tools, data science notebooks, and more. 

UDFs are particularly valuable for sharing logic across multiple tools, standardizing complex business calculations, improving performance for compute-intensive operations (since they're compiled and optimized by your warehouse's query engine), and version controlling custom logic within your dbt project.
```

QMalcolm replied:
The suggestion seems technically accurate 🙂

graciegoheen replied:
Looks good to me - my one suggestion would be to say "They are objects in the warehouse" instead of "They run natively in the warehouse"

runleonarun reviewed (approved):
@luna-bianca approved with some suggestions!

luna-bianca committed:
Update udfs.md

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,206 @@
+
+- BigQuery
+- Snowflake
+- Redshift
```
Databricks as well 🚀 
```suggestion
- Redshift
- Databricks
```

QMalcolm commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,206 @@
+- BigQuery
+- Snowflake
+- Redshift
+- Postgres
```
Databricks is now also supported 🚀 
```suggestion
- Postgres
- Databricks
```

QMalcolm reviewed (approved):
Some minor suggestions but otherwise, looking fantastic! 🚀

runleonarun committed:
Apply suggestions from code review

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

runleonarun committed:
Apply suggestions from code review

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+UDFs are not yet supported in the dbt Fusion engine.
+:::
+
+User-defined functions (UDFs) enable users to define and register custom functions in your warehouse. Like [macros](/docs/build/jinja-macros), UDFs promote code reuse. They run natively in the warehouse so you can reuse the same logic in tools outside dbt.
```
eventually, we will also support UDAFs and UDTFs - not sure if we want to generally refer to these as "functions" or "UDFs"

runleonarun replied:
Maybe functions in this intro paragraph, but going forward we frame it as "we have two kinds of UDFs: UDAFs and UDTFs"

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+User-defined functions (UDFs) enable users to define and register custom functions in your warehouse. Like [macros](/docs/build/jinja-macros), UDFs promote code reuse. They run natively in the warehouse so you can reuse the same logic in tools outside dbt.
+
+dbt creates, updates, and renames UDFs as part of DAG execution. The UDF file is created before building the model that references it.
```
```suggestion
dbt creates, updates, and renames UDFs as part of DAG execution. The UDF is built in the warehouse before the model that references it.
```

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+id: "udfs"
+---
+
+# User-defined functions <Lifecycle status="beta" />
```
must be on `1.11` or `latest` or `fusion` - do we want to call that out anywhere?

runleonarun replied:
Yes! We should add a prereqs section.

runleonarun replied:
Added prereqs [here](https://github.com/dbt-labs/docs.getdbt.com/pull/7841/commits/c0e13d022f620b81809b1e0ac2cce7829cd69278)

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+    - `scalar` - Returns a single value per row
+    - `aggregate` - Returns a single value per group, aggregating several rows
+    - `table functions` - Returns a table result
```
```suggestion
    - `table` - Returns a table result
```

luna-bianca replied:
> 

Hi @graciegoheen , just to note, this expandable is currently hidden because `aggregate` and `table` types are not yet supported.

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+    ```
+    </File>
+
+    In your DAG, there should be a dependency between `is_positive_int` → `my_model` and a UDF node is created from the SQL and YAML definitions.
````
```suggestion
    In your DAG, a UDF node is created from the SQL and YAML definitions, and there will be a dependency between `is_positive_int` → `my_model`.
```

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+    maybe_positive_int_column,
+	udf_db.udf_schema.is_positive_int(maybe_positive_int_column) as is_positive
+
+    from analytics.<dbt_schema>.a_model_i_like
```
```suggestion
    from analytics.dbt_schema.a_model_i_like
```

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+    In your DAG, there should be a dependency between `is_positive_int` → `my_model` and a UDF node is created from the SQL and YAML definitions.
+
+After defining a UDF, you can update the SQL file that contains its function body (`is_positive_int.sql` in this example). When you rebuild, your changes will be applied everywhere the UDF is referenced. 
```
```suggestion
After defining a UDF, if you update the SQL file that contains its function body (`is_positive_int.sql` in this example) or its configurations, your changes will be applied to the UDF in the warehouse next time you `build`. 
```

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+To select UDFs when running a project, use the following commands:
+
+- `dbt run --resource_type function` &mdash; Use this command to only run UDFs in your project.
+- `dbt run --select resource_type:function` &mdash; Use this command to reinitialize all UDFs in your project.
```
@QMalcolm does both `--select resource_type:function` and `--resource_type function` work for `build`? i thought `--resource-type` was just for `list`?

https://docs.getdbt.com/reference/node-selection/methods#resource_type

```suggestion
- `dbt build --select "resource_type:function"` &mdash; Use this command to build all UDFs in your project.
```

QMalcolm replied:
`--resource-type` _does_ work for build 🤔 Just tested it both in cloud and in core. Of note the documentation says `--resource_type` and it should be `--resource-type`

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+To list UDFs in your project, run `dbt list`. 
+
+To select UDFs when running a project, use the following commands:
```
```suggestion
To select UDFs when building a project, use the following commands:
```

"run" is just for models :)

graciegoheen replied:
For the selector examples, we could just include the basic `dbt build --select "resource_type:functions"` and then put other selector examples here -> https://docs.getdbt.com/reference/node-selection/methods#resource_type

luna-bianca replied:
> For the selector examples, we could just include the basic `dbt build --select "resource_type:functions"` and then put other selector examples here -> https://docs.getdbt.com/reference/node-selection/methods#resource_type

Updated the Node selector methods page [here](https://github.com/dbt-labs/docs.getdbt.com/pull/7841/files#diff-3310fdad7b8c50eb9157ee9c418dfda945b38b16b2721a97e167689cac1982e2)

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+## Listing and selecting UDFs
+
+To list UDFs in your project, run `dbt list`. 
```
probably want to add a selector? otherwise `dbt list` will just list all nodes in your project

`dbt list --select "resource_type:function"`

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+- `dbt run --resource_type function` &mdash; Use this command to only run UDFs in your project.
+- `dbt run --select resource_type:function` &mdash; Use this command to reinitialize all UDFs in your project.
+- `dbt run --select path/to/my_function.sql` &mdash; Use this command to select a function by file path.
```
```suggestion
- `dbt build --select path/to/my_function.sql` &mdash; Use this selector to build function by file path.
```

this is the same command - always `dbt build` - your examples are just of different selectors

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+- `dbt run --resource_type function` &mdash; Use this command to only run UDFs in your project.
+- `dbt run --select resource_type:function` &mdash; Use this command to reinitialize all UDFs in your project.
+- `dbt run --select path/to/my_function.sql` &mdash; Use this command to select a function by file path.
+- `dbt run --select my_function` &mdash; Use this command if you modified a UDF and you want to replace it in the data warehouse. To update all models that use the UDF, run `dbt run --select my_function+`.
```
```suggestion
- `dbt build --select my_function` &mdash; Use this selector to build `my_function`. To build all models that use the UDF, run `dbt build --select my_function+`.
```

graciegoheen reviewed (commented):
This is a good start - left some feedback / suggestions.

For the selector examples, we could just include the basic `dbt build --select "resource_type:functions"` and then put other selector examples here -> https://docs.getdbt.com/reference/node-selection/methods#resource_type

Other pages we will want to add:
- **reference docs** -> general spec and resource configs / properties "for functions" (functions are a new node type)
<img width="246" height="537" alt="Screenshot 2025-10-06 at 6 31 53 PM" src="https://github.com/user-attachments/assets/842e469b-443d-466c-9713-ecfbca9fcc5a" />

```
functions:
  - name: string # Required
    description: string # Optional
    type: scalar # Optional, defaults to scalar. Will eventually also support aggregate and table 
    config: # Optional
      # ... inherits from NodeConfig so has things like alias/schema/database/tags/etc
    arguments: # Optional
      - name: string # Required
        data_type: string # Required
        description: string # Optional
    returns: # Required
      data_type: string # Required
      description: string # Optional
```
- nested pages for each of those function-specific properties 
  - for now, i believe this is just `type`, `returns`, `arguments` 
  - on `returns` & `arguments` page, we should call out that `data_type` is warehouse-specific
  - eventually this will include `volatility` and others 
- perhaps an [FAQ](https://docs.getdbt.com/docs/faqs) on when to use UDFs vs. a macro

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+        description: My UDF that determines if a string represents a positive (+) integer # required
+        config:
+            schema: udf_schema
+            database: udf_db
```
```suggestion
          schema: udf_schema
          database: udf_db
```

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+        arguments: # optional
+            - name: a_string # required if arguments is specified
+              data_type: string # required if arguments is specified
+              description: The string that I want to check if it's representing a positive integer (like "10") 
```
```suggestion
          - name: a_string # required if arguments is specified
            data_type: string # required if arguments is specified
            description: The string that I want to check if it's representing a positive integer (like "10") 
```

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+              data_type: string # required if arguments is specified
+              description: The string that I want to check if it's representing a positive integer (like "10") 
+        returns: # required
+            data_type: boolean # required
```
```suggestion
          data_type: boolean # required
```

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+    ```
+    </TabItem>
+
+    <TabItem value="Redshift">
+    ```sql
+    CREATE OR REPLACE FUNCTION repeat_n(s VARCHAR, n INTEGER)
+    RETURNS VARCHAR
+    VOLATILE
+    AS $$ SELECT repeat(s, n); $$
+    LANGUAGE SQL;
````
@QMalcolm this doesn't look like the same function as the snowflake example!!

QMalcolm replied:
In redshift the end result will be
```sql
 CREATE OR REPLACE FUNCTION repeat_n(VARCHAR, INTEGER)
    RETURNS VARCHAR
    VOLATILE
    AS $$
      SELECT repeat($1, $2)
    $$
    LANGUAGE SQL;
```

QMalcolm replied:
If we want it to be the same function it'd beeee
```sql
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(VARCHAR)
    RETURNS BOOLEAN
    VOLATILE # Technically this function could be set as STABLE, but we don't support setting volatility yet
    AS $$
      REGEXP_CONTAINS($1, r'^[0-9]+$')
    $$ LANGUAGE SQL;
```

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+
+    <TabItem value="Snowflake">
+    ```sql
+    CREATE OR REPLACE FUNCTION my_schema.is_positive_int(a_string STRING)
````
from the yaml above, shouldn't this be:
```suggestion
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
```

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+    ```sql
+    {{ config(
+        database='udf_db'
+    ) }}
````
this is duplicative of the database config in yml 

```suggestion
```

graciegoheen replied:
though maybe we want to callout that like models you can specify configs in the .sql file _or_ in yml?

QMalcolm commented:
> @QMalcolm why is Postgres supported locally but not in platform?
> <img alt="Screenshot 2025-10-06 at 6 35 37 PM" width="516" height="399" src="https://private-user-images.githubusercontent.com/53586774/498054268-83429c49-6b2f-45a3-be55-afc255377ec2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk3OTQyNjAsIm5iZiI6MTc1OTc5Mzk2MCwicGF0aCI6Ii81MzU4Njc3NC80OTgwNTQyNjgtODM0MjljNDktNmIyZi00NWEzLWJlNTUtYWZjMjU1Mzc3ZWMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA2VDIzMzkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0ZDgwMmUwNjkwMmJmNGEyMmJjOWYyNDM1YzY5OThhMWE0NjkyYzI2MTlhMmU4YjIzNGRlOGE3YWI0ZjQ1MjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EBRfBy6M7SiJ8VUkb_lot-piXaKUtASAy2JX75YrtAs"> <img alt="Screenshot 2025-10-06 at 6 35 44 PM" width="551" height="366" src="https://private-user-images.githubusercontent.com/53586774/498054296-4bddc266-ecda-4b4a-92d4-6cf468f5d8a9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk3OTQyNjAsIm5iZiI6MTc1OTc5Mzk2MCwicGF0aCI6Ii81MzU4Njc3NC80OTgwNTQyOTYtNGJkZGMyNjYtZWNkYS00YjRhLTkyZDQtNmNmNDY4ZjVkOGE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA2VDIzMzkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhZWM1NWNkOWEzYzVlODFjMzgzMzVlMDRlZDA1MTU0NDk0ODlhNTM4ZmZiNzg5N2MzOWJjZWE4Y2YxNGVmZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IBJm1yd_y2WpwVWm58NfRjL9zKxhxAlIIFTJWKvXXT4">

It's supported with core cloud, but my understanding is that we don't support postgres with fusion.

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+    ```sql
+    select
+
+    maybe_positive_int_column,
+        {{ function('is_positive_int') }}(maybe_positive_int_column)
+
+    from {{ ref('a_model_i_like') }}
````
```suggestion
    select
        maybe_positive_int_column,
        {{ function('is_positive_int') }}(maybe_positive_int_column)
    from {{ ref('a_model_i_like') }}
```

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+    <File name="models/my_model.sql">
+
+    ```sql
+    select
+
+    maybe_positive_int_column,
+	udf_db.udf_schema.is_positive_int(maybe_positive_int_column) as is_positive
+
````
```suggestion
    select
        maybe_positive_int_column,
	    udf_db.udf_schema.is_positive_int(maybe_positive_int_column) as is_positive
```

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+    ```
+    </File>
+
+    In your DAG, there should be a dependency between `is_positive_int` → `my_model` and a UDF node is created from the SQL and YAML definitions.
````
do we have a screenshot of lineage we could add here? to show visualization of the DAG?

QMalcolm replied:
Here is a lineage example 
<img width="1464" height="903" alt="Screenshot 2025-10-06 at 20 36 16" src="https://github.com/user-attachments/assets/04381137-71b7-4a92-8044-33abdd9d33d2" />

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+You can use [unit tests](/docs/build/unit-tests) to validate models that reference UDFs. Before running unit tests, make sure the function exists in your warehouse. To ensure that the function exists for a unit test, run:
+
+```bash
+dbt build --select "+my_model_to_test" --empty
````
@QMalcolm does `--empty` work with functions? that doesn't make sense to me

QMalcolm replied:
@graciegoheen Empty doesn't do _anything_ to functions. However, most unit tests also reference _other models_. For unit tests that depend on other models, the upstream models need to exists (so we can get some information about them). Because of that, we recommend `dbt run --select "+my_model_to_test" --empty` being run first so that the upsteam models exist (but aren't costly in insuring their existence). Doing the `build` in the documentation will also build the upstream models, and thus we should still recommend `--empty`

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+<Tabs>
+
+<TabItem value="core" label="dbt Core local install">
```
```suggestion
<TabItem value="core" label="dbt Core">
```

sounds like this is the list for Core - local and in the platform

graciegoheen commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,208 @@
+
+</TabItem>
+
+<TabItem value ="dbt platform" label ="dbt Core in the dbt platform">
```
```suggestion
<TabItem value ="fusion" label ="dbt Fusion">
```

sounds like this is the list for fusion - local and in the platform

graciegoheen reviewed (dismissed):
Made some more comments / suggestions - I don't think this is ready to merge yet

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+
+    <TabItem value="Redshift">
+    ```sql
+    CREATE OR REPLACE FUNCTION repeat_n(s VARCHAR, n INTEGER)
+    RETURNS VARCHAR
+    VOLATILE
+    AS $$ SELECT repeat(s, n); $$
+    LANGUAGE SQL;
````
to take a guess at syntax looking at redshift [docs](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_FUNCTION.html) (i HAVE NOT VALIDATED)

```suggestion
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
    RETURNS BOOLEAN
    STABLE
    AS $$ SELECT REGEXP_CONTAINS(a_string, r'^[0-9]+$'); $$
    LANGUAGE SQL;
```

things i don't know:
- does redshift adapter automatically inject the `SELECT` or do you need to add that to the .sql file?
- i think volatility is stable for this example? though maybe it's `None` since we don't support `volatility` config yet?

graciegoheen replied:
do we want docs for `volatility` as well? 

`volatility` - “behavior of the function when returning results”:
- `deterministic` functions, like `substr(..)` return the same output for the same input.
- `stable` functions, like `now()` return the same output within one query execution.
- `non-deterministic` functions, like `first()`, might return a different output for the same input.

core issue here -> https://github.com/dbt-labs/dbt-core/issues/12030

By default the volatility of a function is `None`

QMalcolm replied:
Volatility isn't set-able yet. It's hard coded in redshift because we have to supply a volitility redshift (it doesn't let you not)

graciegoheen commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,208 @@
+    LANGUAGE SQL;
+    ```
+    </TabItem>
+    </Tabs>
````
can we get examples of the syntax for the other supported warehouses as well please?

QMalcolm replied:
### Postgres
```
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
    RETURNS BOOLEAN
    AS $$
      REGEXP_CONTAINS(a_string, r'^[0-9]+$')
    $$ LANGUAGE SQL;
```

QMalcolm replied:
### BigQuery
```
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
    RETURNS BOOLEAN
    AS (
      REGEXP_CONTAINS(a_string, r'^[0-9]+$')
    );
```

QMalcolm replied:
### Databricks
```
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
    RETURNS BOOLEAN
    LANGUAGE SQL
    RETURN
      REGEXP_CONTAINS(a_string, r'^[0-9]+$');
```

runleonarun committed:
Apply suggestions from code review

Co-authored-by: Grace Goheen <53586774+graciegoheen@users.noreply.github.com>

runleonarun committed:
Apply suggestion from @runleonarun

runleonarun committed:
Revise UDFs documentation with prerequisites

Updated UDFs documentation to include prerequisites and removed unsupported engine note.

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,200 @@
+
+## Prerequisites
+
+* Make sure you're using the dbt Fusion engine or dbt Core v1.11.
```
```suggestion
* Make sure you're using dbt Core v1.11 or the dbt Fusion engine or Latest in the dbt platform.
```

runleonarun committed:
Apply suggestions from code review

runleonarun committed:
Update website/docs/docs/build/udfs.md

Co-authored-by: Grace Goheen <53586774+graciegoheen@users.noreply.github.com>

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,200 @@
+
+## Prerequisites
+
+* Make sure you're using dbt Core v1.11 or the dbt Fusion engine or Latest in the dbt platform.
```
```suggestion
* Make sure you're using dbt Fusion engine or Latest release track in the dbt Platform or dbt Core v1.11.
```

runleonarun committed:
Update website/docs/docs/build/udfs.md

runleonarun commented on website/docs/docs/build/udfs.md:
```diff
@@ -0,0 +1,200 @@
+
+## Prerequisites
+
+* Make sure you're using dbt Fusion engine or Latest release track in the dbt Platform or dbt Core v1.11.
```
```suggestion
* Make sure you're using dbt Platform's **Latest Fusion** or **Latest** [release track](/docs/dbt-versions/cloud-release-tracks) or dbt Core v1.11.
```

runleonarun committed:
Apply suggestion from @runleonarun

runleonarun commented:
@QMalcolm We could use your help in answering these questions:
* [@QMalcolm does both --select resource_type:function and --resource_type function work for build? i thought --resource-type was just for list?](https://github.com/dbt-labs/docs.getdbt.com/pull/7841/files#r2408960473)
* [@QMalcolm this doesn't look like the same function as the snowflake example!!](https://github.com/dbt-labs/docs.getdbt.com/pull/7841#discussion_r2408973653)
* [Syntax for other supported WHs](https://github.com/dbt-labs/docs.getdbt.com/pull/7841#discussion_r2408993212)

QMalcolm commented on website/docs/docs/build/udfs.md:
````diff
@@ -0,0 +1,200 @@
+    You can use these values for the `type` property when you define a function in a YAML file.
+
+    - `scalar` - Returns a single value per row
+    - `aggregate` - Returns a single value per group, aggregating several rows
+    - `table` - Returns a table result
+    <br></br>
+    For example:
+
+    ```yml
+    functions:
+	  - name: string 
+	    description: string
+	    type: scalar # default value
+    ```
+
+    If not explicitly specified, the `type` property defaults to `scalar`.
+
+    </Expandable>
+    -->
+    
+    The rendered `CREATE` UDF statement depends on which adapter you’re using. For example:
+
+    <Tabs>
+
+    <TabItem value="Snowflake">
+    ```sql
+    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
+    RETURNS BOOLEAN
+    AS (
+    REGEXP_CONTAINS(a_string, r'^[0-9]+$')
+    );
+    ```
````
Should update to:
```suggestion
    ```sql
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
    RETURNS BOOLEAN
    LANGUAGE SQL
    AS $$
      REGEXP_CONTAINS(a_string, r'^[0-9]+$')
    $$;
    ```
```

luna-bianca committed:
Apply changes from comments

luna-bianca committed:
Merge branch 'current' into add-UDFs-doc

runleonarun committed:
image DAG

runleonarun commented:
@luna-bianca here's the follow-up on the rest of the TODOs:

✅ Added screenshot of DAG in [this commit](https://github.com/dbt-labs/docs.getdbt.com/pull/7841/commits/b1d7004107c1f134329a30ee35333d036a16850b)
❌ Volatility doc - not needed until after Coalesce
✅ Drafted reference content in [this PR ](https://github.com/dbt-labs/docs.getdbt.com/pull/8005)

Note that the reference PR is branched off of this PR and I will do the same with the [upgrade guide](https://github.com/dbt-labs/docs.getdbt.com/pull/7998) so links will work and we can merge together.

runleonarun committed:
Merge branch 'current' into add-UDFs-doc

luna-bianca committed:
Replace screenshot

runleonarun committed:
Merge branch 'current' into add-UDFs-doc

mirnawong1 committed:
Merge branch 'current' into add-UDFs-doc

graciegoheen commented on website/static/img/docs/building-a-dbt-project/UDF-DAG.png:
we have an updated lineage view in VSCE! let's use a screenshot of that instead

runleonarun replied:
Added!

runleonarun committed:
Core beta v1.11 upgrade guide (#7998)

## What are you changing in this pull request and why?

Drafting Core 1.11 upgrade guide for Monday. 

## Checklist
<!-- Ensure your PR meets the following items. Feel free to add
additiona checklist items to the list if additional checks need to
happen before the PR is merged, such as:
- [ ] Needs technical review
- [ ] Change base branch
- [ ] Merge on xyz date
-->
- [ ] The changes in this PR meet the [docs style
guide/fundamentals](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md)
required for all content.
- [ ] Applied the proper versioning rules if the content is for specific
dbt version(s): ([version a whole
page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version)
or [version a block of
content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content)
- [ ] The content in this PR requires a dbt release note, so I added one
to the [release notes
page](https://docs.getdbt.com/docs/dbt-versions/dbt-cloud-release-notes).).
<!--
PRE-RELEASE VERSION OF dbt (if so, uncomment):
- [ ] Add a note to the prerelease version [Migration
Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
<!-- 
ADDING OR REMOVING PAGES (if so, uncomment):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/vercel.json`
- [ ] Run link testing locally with `npm run build` to update the links
that point to deleted pages
-->

---------

Co-authored-by: Bianca Luna <biancacamille.luna@dbtlabs.com>
Co-authored-by: Grace Goheen <53586774+graciegoheen@users.noreply.github.com>

runleonarun committed:
Adding references for Functions (#8005)

## What are you changing in this pull request and why?

This PR adds the follow up reference content for
https://github.com/dbt-labs/docs.getdbt.com/pull/7841

* Added an FAQ (`website/docs/faqs/Project/udfs-vs-macros.md`) to help
users decide when to use UDFs versus macros, with practical scenarios
and examples showing how both can be combined.
* Updated function configuration reference
(`website/docs/reference/function-configs.md`) to describe available
configuration options for UDFs, hierarchical config precedence, and
example YAML setups for project-wide, directory-level, and
single-function configs.
* Added function properties reference
(`website/docs/reference/function-properties.md`) detailing the YAML
structure for declaring UDFs, including required and optional fields,
with a full example.
* Updated resource type documentation to include `function` as a
selectable resource type in dbt v1.11 and above
* Enhanced node selection method docs to show how UDFs can be selected
by file and path, with examples for both models and functions for dbt
v1.11+.

## Checklist
<!-- Ensure your PR meets the following items. Feel free to add
additiona checklist items to the list if additional checks need to
happen before the PR is merged, such as:
- [ ] Needs technical review
- [ ] Change base branch
- [ ] Merge on xyz date
-->
- [ ] The changes in this PR meet the [docs style
guide/fundamentals](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md)
required for all content.
- [ ] Applied the proper versioning rules if the content is for specific
dbt version(s): ([version a whole
page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version)
or [version a block of
content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content)
- [ ] The content in this PR requires a dbt release note, so I added one
to the [release notes
page](https://docs.getdbt.com/docs/dbt-versions/dbt-cloud-release-notes).).
<!--
PRE-RELEASE VERSION OF dbt (if so, uncomment):
- [ ] Add a note to the prerelease version [Migration
Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
<!-- 
ADDING OR REMOVING PAGES (if so, uncomment):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/vercel.json`
- [ ] Run link testing locally with `npm run build` to update the links
that point to deleted pages
-->

<!-- vercel-deployment-preview -->
---
🚀 Deployment available! Here are the direct links to the updated files:


-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/docs/build/udfs
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/faqs/Project/udfs-vs-macros
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/dbt_project.yml
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/function-configs
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/function-properties
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/project-configs/function-paths
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/resource-properties/arguments
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/resource-properties/function-arguments
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/resource-properties/returns
-
https://docs-getdbt-com-git-udf-refs-dbt-labs.vercel.app/reference/resource-properties/type

<!-- end-vercel-deployment-preview -->

---------

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>
Co-authored-by: Grace Goheen <53586774+graciegoheen@users.noreply.github.com>
Co-authored-by: mirnawong1 <mirnawong@gmail.com>

runleonarun committed:
adding new image and updating FAQ

runleonarun committed:
Merge branch 'current' into add-UDFs-doc
